### PR TITLE
Fix skopeo layers and deprecate it

### PR DESCRIPTION
**Pull request by runcom**
As title says. @mtrmac PTAL

@giuseppe I know you're using `skopeo layers` in your sys containers, would be nice to migrate that to `mkdir dir && skopeo copy docker://image_name dir:here_are_the_layers`

### CODE REVIEW

File: cmd/skopeo/layers.go [hunk 1]
```diff
@@ -15,10 +17,14 @@ import (
 var layersCmd = cli.Command{
 	Name:      "layers",
 	Usage:     "Get layers of IMAGE-NAME",
-	ArgsUsage: "IMAGE-NAME",
+	ArgsUsage: "IMAGE-NAME [LAYER...]",
 	Action: func(c *cli.Context) error {
+		fmt.Println(`DEPRECATED: skopeo layers is deprecated in favor of skopeo copy `)
```
Comment (mtrmac) [inline cmd/skopeo/layers.go]: Shouldn’t this go to stderr? I guess I can see a case for either… just something to consider.

File: cmd/skopeo/layers.go [hunk 2]
```diff
@@ -15,10 +17,14 @@ import (
...
+		if c.NArg() == 0 {
+			return errors.New("please specify an image")
+		}
```
Comment (mtrmac) [inline cmd/skopeo/layers.go]: ACK on having the check; for consistency, this should say `usage: correct usage`, like e.g. https://github.com/projectatomic/skopeo/blob/master/cmd/skopeo/delete.go#L12

**Comment by mtrmac** [comment]
> would be nice to migrate that to `mkdir dir && skopeo copy docker://image_name dir:here_are_the_layers`

(Do we document and commit to the `dir:` format as an ABI? I guess we need to… Not that having `skopeo layers` undocumented ever stopped anyone 😈 )

**Comment by runcom** [comment]
> (Do we document and commit to the dir: format as an ABI? I guess we need to… Not that having skopeo layers undocumented ever stopped anyone  )

I think we should commit to `oci:` instead here

**Comment by mtrmac** [comment]
> I think we should commit to `oci:` instead here

That might involve a schema or metadata conversion… which might actually be desirable, depending on the situation.

**Comment by runcom** [comment]
alright, fixed your comments - we'll talk about atomic moving to copy in another issue maybe (no urgency right now)

**Comment by mtrmac** [comment]
Thanks! LGTM pending tests.

**Comment by rhatdan** [comment]
Lets remove the layers information in the man page.  Perhaps remove layers from command completion.  I would prefer we make it dissapear if we don't want users using it any more.

**Comment by giuseppe** [comment]
one advantage of `skopeo layers` is that I could specify which layers to download, so that `atomic pull` will only fetch the missing ones (a layer could be already stored locally, from another image, or didn't change in a newer version of the image).  How can I do it with `copy`?

**Comment by mtrmac** [comment]
`copy` can’t do that for `dir:`; it would be possible to write an `ostree:` transport for containers/image and then `skopeo copy` would be able to skip redownloading, but that’s obviously not all that attractive given the existing Python implementation.

**Comment by giuseppe** [comment]
could we add a filter to `skopeo copy` that allows to specify which layers to consider as `skopeo layers` currently does?  We can drive it from atomic in the same way as we currently do.

**Comment by mtrmac** [comment]
In principle, sure, it’s code. Though it is a bit unclear what this does to the semantics of `skopeo copy` (will the manifest still contain all layers?), and some cases (schema conversion) really need the layers at least read if not written. So, it would be pretty ugly in an already pretty complex copy pipeline.

Separately, what semantics exactly do you need? Is it just a blind “download one layer, the caller is reading and interpreting the manifest and dealing with any semantics”, or it the caller actually expecting intelligence and conversions etc. from skopeo?

**In particular, should signatures be verified when copying the manifest and data** for your use case?

**Comment by giuseppe** [comment]
what we do in atomic is to first fetch the manifest, parse it, and then retrieve all the layers that are not already stored in the OSTree repository.  What is needed from atomic is the possibility to fetch a subset of the layers in an image.
The only intelligence needed is for the signatures.  Are signatures verified on the manifest?

**Comment by mtrmac** [comment]
Right now, `skopeo inspect` does not enforce the signature policy, only `skopeo copy` does. It could be added to `inspect`, at least as an option.  After that, whatever is downloading thes layer *must* verify the layer blobs against their digests, which current `skopeo layers` isn’t doing either.

@runcom Opinions on the path forward?

It seems to me that we either ignore signatures, in which case it seems most natural to revive `skopeo layers`, Or we plan to support them for pulling to `ostree:`, in which case a native Golang `ostree:` transport would be conceptually simpler than carving up `skopeo copy` into pieces and then having `atomic` reassemble it back together. OTOH https://github.com/projectatomic/atomic/blob/master/Atomic/syscontainers.py is pretty big.

… after all, a `skopeo copy` restricted to only a smaller set of layers starts looking as a pretty attractive option…

**Comment by giuseppe** [comment]
would it be possible to check signatures having only a subset of the layers?  Having something like `--skip-layers=....` seems to me like a quite generic solution and won't require to change what we already have in place.

**Comment by rhatdan** [comment]
In the future we would also want to support container/storage as a back end for skopy copy also.  Is there a simple way to get that to happen also?

**Comment by mtrmac** [comment]
containers/storage will be integrated natively ( https://github.com/containers/image/pull/63 ), so these disassemble/reassemble-at-the-CLI-level concerns do not apply.  (Also, it _always_ reads all inputs in https://github.com/containers/image/pull/63/files#diff-d77114c36cb773f7266e73a470816298R218 so it is not as interested in skipping existing layers.)

**Comment by mtrmac** [comment]
> … after all, a `skopeo copy` restricted to only a smaller set of layers starts looking as a pretty attractive option…

> would it be possible to check signatures having only a subset of the layers? Having something like --skip-layers=.... seems to me like a quite generic solution and won't require to change what we already have in place.

Well, “generic” only to the `atomic`-like “dig into the internal implementation details” users. But yes, it does seem the simplest way to integrate the two; https://github.com/containers/image/pull/159 is not _pretty_ but it seems fairly acceptable, at least if we add some tests… @runcom ?

**Comment by rhatdan** [comment]
@mtrmac Are you saying that there will be a 
skopeo copy docker:image storage
So that atomic command could pull an image from a registry and install it directly into a local container/storage repository?

**Comment by mtrmac** [comment]
> Are you saying that there will be a skopeo copy docker:image storage
> So that atomic command could pull an image from a registry and install it directly into a local container/storage repository? 

I’d expect `skopeo copy docker:busybox storage:busybox` to work with https://github.com/containers/image/pull/63 , already today, yes.

**Comment by rhatdan** [comment]
Awesome, if this was merged...

**Pull request merged**